docs/proposals/env-vars.md

```diff
@@ -0,0 +1,565 @@
+us to pull content from `Secrets`, like authentication token.  Many of the tools when building images (including
+source code extraction and tools that have to download dependencies) consume authentication tokens via environment
+variables.  Storing those token in k8s `Secrets` vs. having to specify them in scripts referenced in `Containers` has 
+become a preferred method when building images in k8s native environments.
+
+Next, the question of whether to use the k8s `EnvVar` directly exists.  This proposal suggests creating a new
+Shipwright type that inlines the k8s type, in case we want to extend and add new features later that require
+addtiional fields or new types from k8s or Tekton.  So with that, we start with
+
+```go
+type EnvironmentVariable struct {
+   corev1.EnvVar
+   
+}
+```
+
+NOTE: `valueFrom` in k8s has a range of options, from `Secret` refs, `ConfigMap` refs, field refs, etc.  The `Secret` ref
+is the one we see immediate need for, but the EP does *NOT* proscribe restrictions at this time.  Of course, as we progress
+toward implementation and customer feedback, if need be, restrictions can be added at a later date.
+
+Then, let's examine the updates for `BuildStrategy` (both cluster and namespaced scope), `Build`, and `BuildRun` and where we
+might update them with an array of `EnvironmentVariable`
+
+`BuildStrategy` needs some further consideration, in that `BuildStep` already inlines `Container`, which allows specification 
+of `EnvVar`.
+
+Those existing `BuildStep` / `Container` environment variables can be the way an administrator could ensure that 
+any environment variable exists and has a given setting.  These would be environment variables that could not be 
+changed by a `Build` or `BuildRun`.
+
+But in the case where an organization wants to provide a default environment variable setting, but allow their `Build` and
+`BuildRun` creators to override the value for that setting (including setting it to the empty string, which in most 
+cases would mean "turning off" the environment variable), the `BuildStrategySpec` would have an array of Shipwright
+`EnvironmentVariable` added to it.
+
+Lastly, so there are two basic mindsets from an admin perspective:
+
+- Allow any environment variable in general, but restrict a few
+- Disallow any environment variable in general, but allow a few
+
+To facilitate both mindsets, we'll add two string arrays to the `BuildStrategySpec`
+- `AllowedEnvironmentVariables`
+- `DisallowedEnvironmentVariables`
+
+Each string array list the `EnvVar` names/keys to consider.
+ 
+If `AllowedEnvironmentVariables` is empty, assume all are allowed, except for those mentioned in `DisallowedEnvironmentVariables`.  
+Otherwise, the list is the list, and any environment variables in `BuildStep`, `Build`, and `BuildRun` must be vetted against this list.  
+Initially, no wildcards or regexp, but perhaps a feature to consider in the future.
```

Comment:
I think you can remove this line as you later mention

> Each entry of the allow and disallow arrays need to minimally support wildcards, if not regular expressions.

Reply:
agreed